### PR TITLE
Core modules split

### PR DESCRIPTION
**Pull request by KTGypsy**
## What:

* Project split into core and operations-api.
* Some minor workflows modification, detekt -> detektAll
* Gradle scripts 

## Why:

* To introduce architecture which allows creating plugins with additional operations. By using only operations-api module user is able to create new library and attach operations from it during engine initialisation process.

## Example:

* Internal usage of operations-api to create stdlib module: https://github.com/allegro/json-logic-kmp/pull/41

### CODE REVIEW

File: operations-api/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import org.jetbrains.kotlin.gradle.plugin.mpp.apple.XCFramework
+
+plugins {
+    kotlin("multiplatform") version Versions.kotlin
+    id("io.kotest.multiplatform") version Versions.kotest apply false
```
Comment (KTGypsy) [inline operations-api/build.gradle.kts:5]: Removing this plugin causes:
Cannot add task 'commonizeNativeDistribution' as a task with that name already exists.
So I disabled it with false parameter

Comment (KTGypsy) [reply operations-api/build.gradle.kts:5]: Any ideas how to solve it in other way?

File: core/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+}
+
+/*
+We need to disable this task, because it fails with following error:
+`the path does not point to a valid debug symbols file`.
+Currently we are using only Release XCFramework.
+It could be fixed also by `isStatic = false` but we want to get static lib.
+ */
+tasks.getByName("assembleJsonLogicKMPDebugXCFramework").enabled = false
```
Comment (KTGypsy) [inline core/build.gradle.kts:79]: - [ ] Extract it and apply to every module

**Comment by KTGypsy** [comment]
I removed comments left by reviewdog because of wrong detekt config in the previous commits

**Review by PaulinaSadowskaAllegro — COMMENTED** [review]
FIles I think we should consider adding to core:
- JsonLogicException (because other operations may also want to throw exceptions, right?)
- Array, Boolean and StringUtils? - I may be wrong so we probably should not do it right away, but I think other operations may also need some type checking and parsing

Also: wasn't json logic operators supposed to be in separate module? Now jsonlogic operations are in core 
json-logic-kmp (1)
 (maybe it was wrong assumption and jsonLogic operators should be in core. For DSL moving jsonLogic operations out of core would also generate problems (with autocomplete).

File: operations-api/build.gradle.kts (hunk 1 above)
Comment (KTGypsy) [reply operations-api/build.gradle.kts:5]: Why would an API module know anything about any operators? In the linked diagram this module is independent from any module. Because of its independence there will probably be no concrete implementation/nothing to really test.

**Comment by KTGypsy** [comment]
> FIles I think we should consider adding to core:
> 
> * JsonLogicException (because other operations may also want to throw exceptions, right?)
> * Array, Boolean and StringUtils? - I may be wrong so we probably should not do it right away, but I think other operations may also need some type checking and parsing
> 
> Also: wasn't json logic operators supposed to be in separate module? Now jsonlogic operations are in core json-logic-kmp (1) (maybe it was wrong assumption and jsonLogic operators should be in core. For DSL moving jsonLogic operations out of core would also generate problems (would break autocomplete).

I agree with the first point - I will move the exception to the API module 👍🏽 
Utils are quite controversial because they are here mainly because of how JS works. I would not extract them to the API module because external plugins probably will not need them(?). If we want to really reuse them I propose creating new module with utils/common. By doing so we will not bother new developers with some strange JS specific parsing utils and keep them for internal usage.

Last but not least, the core/contract operators. Solution I prepared is some kind of proposal. I discussed it with Marek to be more sure about my idea. 
What I think is:

- Operations from contract is core functionality which are always included to the engine and there is no way to build one without them.
- All of the operations has been tested with use of the engine so far. If we want to change it we would need to reorganise the way we test them and also new operations from stdlib(for them it is inevitable).
 - (Common)engine builder needs information about standard/contracted operations to add them by default
 
 To sum up from the practical point of view the dependency diagram needs some changes because of the implementation details.

**Review by PaulinaSadowskaAllegro — APPROVED** [review]

**Review by mkrogulskiallegro2 — APPROVED** [review]

**Review by mkrogulskiallegro2 — APPROVED** [review]

**Review by PaulinaSadowskaAllegro — APPROVED** [review]
